### PR TITLE
change MarkingFunctions.NoOp name to MarkingFunctions.Default. Add basic implementations for 'combine' methods

### PR DESCRIPTION
**Pull request by tomnelson**
Changes to be committed:
	modified:   pom.xml
	modified:   services/accumulo-utils/src/main/java/datawave/marking/MarkingFunctions.java
	modified:   warehouse/data-dictionary-core/src/test/java/datawave/webservice/datadictionary/DatawaveDataDictionaryImplTest.java
	modified:   warehouse/ingest-configuration/src/main/resources/MarkingFunctionsContext.xml
	modified:   warehouse/ingest-core/src/main/java/datawave/ingest/config/RawRecordContainerImpl.java
	modified:   warehouse/ingest-core/src/main/java/datawave/ingest/data/config/MarkingsHelper.java
	modified:   warehouse/ingest-core/src/test/resources/MarkingFunctionsContext.xml
	modified:   warehouse/query-core/src/main/java/datawave/query/discovery/DiscoveryTransformer.java
	modified:   warehouse/query-core/src/test/java/datawave/query/IfThisTestFailsThenHitTermsAreBroken.java
	modified:   warehouse/query-core/src/test/java/datawave/query/UseOccurrenceToCountInJexlContextTest.java
	modified:   warehouse/query-core/src/test/java/datawave/query/predicate/ValueToAttributesTest.java
	modified:   warehouse/query-core/src/test/java/datawave/query/tables/IndexQueryLogicTest.java
	modified:   warehouse/query-core/src/test/java/datawave/query/testframework/AbstractFunctionalQuery.java
	modified:   web-services/client/src/main/java/datawave/webservice/response/objects/DefaultKey.java
	modified:   web-services/deploy/configuration/src/main/resources/MarkingFunctionsContext.xml
	modified:   web-services/query/src/test/java/datawave/webservice/query/logic/TestLegacyBaseQueryLogicTransformer.java
	modified:   web-services/query/src/test/java/datawave/webservice/query/logic/composite/CompositeQueryLogicTest.java

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -76,7 +76,7 @@
         <version.log4j>1.2.16</version.log4j>
         <version.log4j-extras>1.0</version.log4j-extras>
         <version.lucene>7.5.0</version.lucene>
-        <version.microservice.accumulo-utils>1.2</version.microservice.accumulo-utils>
+        <version.microservice.accumulo-utils>1.3-SNAPSHOT</version.microservice.accumulo-utils>
         <version.microservice.audit-api>1.4</version.microservice.audit-api>
```
Comment (tomnelson) [inline pom.xml]: this has to be 1.3 (not SNAPSHOT) once the accumulo-utils module is 'released' at 1.3

File: pom.xml [hunk 2]
```diff
@@ -76,7 +76,7 @@
         <version.log4j>1.2.16</version.log4j>
         <version.log4j-extras>1.0</version.log4j-extras>
         <version.lucene>7.5.0</version.lucene>
-        <version.microservice.accumulo-utils>1.2</version.microservice.accumulo-utils>
+        <version.microservice.accumulo-utils>1.3-SNAPSHOT</version.microservice.accumulo-utils>
```
Comment (tomnelson) [inline pom.xml]: this has to be 1.3 (not SNAPSHOT) once the accumulo-utils module is 'released' at 1.3

File: services/accumulo-utils/src/main/java/datawave/marking/MarkingFunctions.java [hunk 3]
```diff
@@ -68,18 +73,29 @@ public Exception(Throwable cause) {
...
+            return new ColumnVisibility(new ColumnVisibility(expressions.stream().filter(viz -> viz.flatten().length > 0)
+                            .map(viz -> "(" + new String(viz.flatten(), UTF_8) + ")").collect(Collectors.joining("&")).getBytes(UTF_8)).flatten());
         }
         
         @Override
         @SafeVarargs
         public final Map<String,String> combine(Map<String,String>... markings) {
-            return markings.length == 0 ? new HashMap<>() : markings[0];
+            Map<String,String> combinedMap = new HashMap<>();
+            Stream.of(markings).flatMap(map -> map.entrySet().stream().filter(entry -> !Strings.isNullOrEmpty(entry.getValue()))).forEach(entry -> {
```
Comment (brianloss) [inline services/accumulo-utils/src/main/java/datawave/marking/MarkingFunctions.java]: My guess is you probably want to operate only on the COLUMN_VISIBILITY entry in the map, if there is one.

File: services/accumulo-utils/src/main/java/datawave/marking/MarkingFunctions.java [hunk 4]
```diff
@@ -13,10 +14,11 @@
 
 import java.io.IOException;
 import java.lang.reflect.InvocationTargetException;
-import java.util.Collection;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.*;
```
Comment (brianloss) [inline services/accumulo-utils/src/main/java/datawave/marking/MarkingFunctions.java]: Can you undo the star imports?

File: services/accumulo-utils/src/main/java/datawave/marking/MarkingFunctions.java [hunk 5]
```diff
@@ -68,18 +70,25 @@ public Exception(Throwable cause) {
...
         public static final String COLUMN_VISIBILITY = "columnVisibility";
         
         @Override
         public ColumnVisibility combine(Collection<ColumnVisibility> expressions) {
-            return expressions.isEmpty() ? new ColumnVisibility() : expressions.iterator().next();
+            
+            // filter out any empty expressions, then flatten each one (to de-dupe) and concatenate with '&'
+            // flatten the final combined ColumnVisibility and use that to make the ColumnVisibility to return
+            return new ColumnVisibility(new ColumnVisibility(expressions.stream().filter(viz -> viz.flatten().length > 0)
+                            .map(viz -> "(" + new String(viz.flatten(), UTF_8) + ")").collect(Collectors.joining("&")).getBytes(UTF_8)).flatten());
```
Comment (brianloss) [inline services/accumulo-utils/src/main/java/datawave/marking/MarkingFunctions.java]: This is calling flatten twice. You could do something like 
```
return new ColumnVisibility(
    new ColumnVisibility(
        expressions.stream()
            .map(ColumnVisibility::flatten)
            .filter(b -> b.length > 0)
            .map(b -> "(" + new String(viz.flatten(), UTF_8) + ")")
            .collect(Collectors.joining("&")).getBytes(UTF_8)
    ).flatten());
```

**Pull request merged**